resource-directory/service/updateResourceValues.go

```diff
@@ -0,0 +1,174 @@
+	switch statusCode {
+	case codes.OK:
+	case extCodes.Accepted:
+	default:
+		s := status.New(statusCode, "response from device")
+		if content != nil {
+			newS, err := s.WithDetails(&errdetails.DeviceError{
+				Content: &errdetails.Content{
+					Data:        content.GetData(),
+					ContentType: content.GetContentType(),
+				},
+			})
+			if err == nil {
+				s = newS
+			}
+		}
+		return nil, s.Err()
+	}
+	return content, nil
+}
+
+func toResponse(processed raEvents.ResourceUpdated) (*pb.UpdateResourceValuesResponse, error) {
+	content, err := eventContentToContent(processed.GetStatus(), processed.GetContent())
+	if err != nil {
+		return nil, err
+	}
+	return &pb.UpdateResourceValuesResponse{
+		Content: content,
+		Status:  pb.RAStatus2Status(processed.GetStatus()),
+	}, nil
+}
+
+func (r *RequestHandler) waitForUpdateContentResponse(ctx context.Context, deviceID, resourceID string, notify <-chan raEvents.ResourceUpdated, onTimeout func(ctx context.Context, destDeviceId, resourceID string, notify <-chan raEvents.ResourceUpdated) (*pb.UpdateResourceValuesResponse, error)) (*pb.UpdateResourceValuesResponse, error) {
+	timeoutCtx, cancel := context.WithTimeout(ctx, r.timeoutForRequests)
+	defer cancel()
+	select {
+	case processed := <-notify:
+		return toResponse(processed)
+	case <-timeoutCtx.Done():
+		return onTimeout(ctx, deviceID, resourceID, notify)
+	}
+}
+
+func (r *RequestHandler) UpdateResourcesValues(ctx context.Context, req *pb.UpdateResourceValuesRequest) (*pb.UpdateResourceValuesResponse, error) {
+	accessToken, err := kitNetGrpc.TokenFromMD(ctx)
+	if err != nil {
+		return nil, logAndReturnError(status.Errorf(codes.Unauthenticated, "cannot update resource: %v", err))
+	}
+	if req.ResourceId == nil {
+		return nil, logAndReturnError(status.Errorf(codes.InvalidArgument, "cannot update resource: invalid ResourceId"))
+	}
+	deviceID := req.GetResourceId().GetDeviceId()
+	href := req.GetResourceId().GetResourceLinkHref()
+	errorMsg := fmt.Sprintf("cannot update resource /%v%v", deviceID, href) + ": %v"
+
+	correlationIDUUID, err := uuid.NewV4()
+	if err != nil {
+		return nil, logAndReturnError(status.Errorf(codes.Internal, errorMsg, err))
+	}
+
+	correlationID := correlationIDUUID.String()
+	resourceID := req.ResourceId.ID()
+	notify := r.updateNotificationContainer.Add(correlationID)
+	defer r.updateNotificationContainer.Remove(correlationID)
+
+	loaded, err := r.resourceProjection.Register(ctx, deviceID)
+	if err != nil {
+		return nil, logAndReturnError(status.Errorf(codes.NotFound, errorMsg, fmt.Errorf("cannot register device to projection: %w", err)))
+	}
+	defer r.resourceProjection.Unregister(deviceID)
+
+	if !loaded {
+		if len(r.resourceProjection.Models(deviceID, resourceID)) == 0 {
+			err = r.resourceProjection.ForceUpdate(ctx, deviceID, resourceID)
+			if err != nil {
+				return nil, logAndReturnError(status.Errorf(codes.NotFound, errorMsg, err))
+			}
+		}
+	}
+
+	connectionID := "grpc-gateway"
```

Comment:
Please make this a constant to avoid repetition

Reply:
done